File: backend/app/api/energy.py
"""Energy log API — spec §9 Today screen energy history."""
from __future__ import annotations

import uuid
from datetime import datetime, timedelta, timezone
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db import get_db
from app.models.energy import EnergyLog


router = APIRouter(prefix="/energy", tags=["energy"])


EnergyLevelLiteral = Literal["low", "medium", "high"]


class EnergyLogOut(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    level: str
    notes: str | None
    logged_at: datetime
    created_at: datetime

    class Config:
        from_attributes = True


class EnergyLogCreate(BaseModel):
    user_id: uuid.UUID
    level: EnergyLevelLiteral
    notes: str | None = None


@router.post("", response_model=EnergyLogOut, status_code=status.HTTP_201_CREATED)
def log_energy(payload: EnergyLogCreate, db: Session = Depends(get_db)) -> EnergyLog:
    row = EnergyLog(
        user_id=payload.user_id,
        level=payload.level,
        notes=payload.notes,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    return row


@router.get("", response_model=list[EnergyLogOut])
def list_energy(
    user_id: uuid.UUID = Query(...),
    days: int = Query(default=14, ge=1, le=365),
    db: Session = Depends(get_db),
) -> list[EnergyLog]:
    since = datetime.now(timezone.utc) - timedelta(days=days)
    stmt = (
        select(EnergyLog)
        .where(EnergyLog.user_id == user_id, EnergyLog.logged_at >= since)
        .order_by(EnergyLog.logged_at.desc())
    )
    return list(db.execute(stmt).scalars())


@router.get("/latest", response_model=EnergyLogOut | None)
def latest_energy(
    user_id: uuid.UUID = Query(...),
    db: Session = Depends(get_db),
) -> EnergyLog | None:
    stmt = (
        select(EnergyLog)
        .where(EnergyLog.user_id == user_id)
        .order_by(EnergyLog.logged_at.desc())
        .limit(1)
    )
    return db.execute(stmt).scalar_one_or_none()

File: mobile/src/api/energy.ts
import { API_BASE_URL } from "../config/api";

export type EnergyLevel = "low" | "medium" | "high";

export interface EnergyLog {
  id: string;
  user_id: string;
  level: EnergyLevel;
  notes: string | null;
  logged_at: string;
  created_at: string;
}

async function handle<T>(resp: Response): Promise<T> {
  if (!resp.ok) {
    let detail = `${resp.status} ${resp.statusText}`;
    try {
      const body = await resp.json();
      if (body?.detail) detail = `${resp.status}: ${JSON.stringify(body.detail)}`;
    } catch {
      /* ignore */
    }
    throw new Error(detail);
  }
  return (await resp.json()) as T;
}

export async function logEnergy(
  userId: string,
  level: EnergyLevel,
  notes?: string
): Promise<EnergyLog> {
  const resp = await fetch(`${API_BASE_URL}/energy`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ user_id: userId, level, notes: notes ?? null }),
  });
  return handle<EnergyLog>(resp);
}

export async function listEnergy(
  userId: string,
  days = 14
): Promise<EnergyLog[]> {
  const qs = new URLSearchParams({
    user_id: userId,
    days: String(days),
  });
  const resp = await fetch(`${API_BASE_URL}/energy?${qs.toString()}`);
  return handle<EnergyLog[]>(resp);
}

export async function latestEnergy(
  userId: string
): Promise<EnergyLog | null> {
  const qs = new URLSearchParams({ user_id: userId });
  const resp = await fetch(`${API_BASE_URL}/energy/latest?${qs.toString()}`);
  return handle<EnergyLog | null>(resp);
}
